examples/particle-swarm-optimizer.dx

```diff
@@ -0,0 +1,101 @@
+' # Particle Swarm Optimizer
+
+' ## Fitness function
+
+rosenbrock:: Real -> Real -> Real
+rosenbrock x y = sq (1.0 - x) + 80.0*sq (y - x*x)
+
+
+' We write one that uses vector for input
+
+rosenbrock2:: (2=>Real) -> Real
+rosenbrock2 xs =
+  x = xs.(asidx 0)
+  y = xs.(asidx 1)
+  rosenbrock x y
+
+
+' Min should be at 1.0, 1.0
+
+:p rosenbrock 1.0 1.000
+
+:p rosenbrock2 [1.0, 1.000]
+
+:p rosenbrock 1.0 1.02
+
+:p rosenbrock2 [1.0, 1.02]
+
+
+
+' ## Helper functions
+
+' make a random vector unifornly distrubuted between lb and ub
+
+randBounded:: Key -> (d=>Real)->(d=>Real)->(d=>Real)
+randBounded key lb ub =
+    for i. lb.i + ((rand $ ixkey key i) * (ub.i - lb.i))
+
+randBounded (newKey 4) [1.0,  -2.0] [-1.0,  2.0]
+
+' minby and minimumby
+' to find the smallest values
+
+minby:: A a::Data. (a->Real)->a->a->a
+minby f x y = [x,y].(asidx $ b2i $ (f x) > (f y))
+
+minby sq -1.0 0.5
+minby (asNonLin fst) (0.7, 8.0) (3.0, 12.0)
+
+
+minimumby:: A p q::Data. (q->Real)->(p=>q)->q
+minimumby f xs = fold xs.(asidx 0) (for i. minby f xs.i)
+
+minimumby sq [-1.0, -0.1, 2.0, 3.0]
+
+
+' ## The Optimizer itself.
+### TODO:
+ - workout how to pass `nparticles` and `niters` in as a variable
+     - Currrently hardcoded as 50 and 140 respectively
+ - workout more compact way to define type sig
+
+
+optimize:: Key->((d=>Real)->Real) -> (d=>Real,d=>Real) -> (Real,Real,Real) -> (d=>Real)
+optimize key f (lb,ub) (momentum,gRate,pRate) =
+    optStep (keyL, gbest, pbests,positions,velocities) =
+        (keyG, keyP, keyNext) = splitKey3 keyL
+        gWeight::Real = gRate * rand keyG
+        pWeight::Real = pRate * rand keyP
```

Comment:
These weights should be different 
(different rand) perr  particle and per dimension.

I am mildly dubious about that being an important part of the algorithm.
But I guess it does increase exploration, and the cost of random number generation is generally cheap vs the fitness function